libraries/c_sdk/standard/https/src/iot_https_client.c

```diff
@@ -1287,17 +1287,21 @@ static void _networkReceiveCallback( void * pNetworkConnection,
 
     IotMutex_Unlock( &( pHttpsConnection->connectionMutex ) );
 
-    /* Signal to a synchronous reponse that the response is complete. */
-    if( pCurrentHttpsResponse->isAsync && pCurrentHttpsResponse->pCallbacks->responseCompleteCallback )
-    {
-        pCurrentHttpsResponse->pCallbacks->responseCompleteCallback( pCurrentHttpsResponse->pUserPrivData, pCurrentHttpsResponse, status, pCurrentHttpsResponse->status );
-    }
-
-    /* For a synchronous request release the semaphore. */
+    /* The first if-case below notifies IotHttpsClient_SendSync() that the response is finished receiving. When
+     * IotHttpsClient_SendSync() returns the user is allowed to modify the user buffer used for the response context.
+     * In the asynchronous case, the responseCompleteCallback notifies the application that the user buffer used for the
+     * response context can be modified. Posting to the respFinishedSem or calling the responseCompleteCallback MUST be
+     * mutually exclusive by wrapping in an if/else. If these were separate if-cases, then there could be a context
+     * switch in between where the application modifies the buffer causing the next if-case to be executed. */
```

Comment:
What the code seems to say, so to speak, is that these are mutually exclusive situations: if you're not async, post to the semaphore. If you are async, try the callback.

Reply:
Yes true.